### ISSUE TITLE
Windows: Renewal deploy hook opens in notepad.exe instead of being executed

### ISSUE THREAD

**Issue by mikenor** [certbot/certbot#7713]
I placed a .ps1 Powershell script into `C:\Certbot\renewal-hooks\deploy` but Certbot does not execute it. Instead, it opens the script with Windows Notepad.

## My operating system is (include version):
Windows 10 1809

## I installed Certbot with (certbot-auto, OS package manager, pip, etc):
https://dl.eff.org/certbot-beta-installer-win32.exe

## I ran this command and it produced this output:
`certbot renew --force-renewal`

```
Saving debug log to C:\Certbot\log\letsencrypt.log

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Processing C:\Certbot\renewal\miketest.pwsd76.ab.ca.conf
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Plugins selected: Authenticator standalone, Installer None
Renewing an existing certificate
Running deploy-hook command: C:\Certbot\renewal-hooks\deploy\LM-CertUpload.ps1
```
Then Windows' notepad.exe opened up, showing my .ps1 script.

## Certbot's behavior differed from what I expected because:
The [forum post about the Certbot Windows beta](https://community.letsencrypt.org/t/beta-phase-of-certbot-for-windows/105822) explicitly states that .ps1 hooks are supposed to work:
> Similarly to Linux, Certbot for Windows supports any file executable for this platform, if it is available in the PATH. It can be .bat, .ps1 or even executable binaries like .exe.

**Comment by adferrand** [comment]
**TL;DR**

Sadly for now Powershell scripts will not work in the `renewal-hooks` folder. Only BAT files or executable binaries will.

If you really want to use Powershell scripts, explicitly pass the executable interpreter of the script to the relevant CLI hook flag. For instance, with a Powershell script, use `--deploy-hook "powershell.exe -File C:\path\to\your\script.ps1"`.

**Complete explanation**

On every OS system (Linux, Windows, ...), you run something by passing a file to one entrypoint of the `exec` family supported by the kernel (`system_call`, `CreateProcess` ...). Only pre-compiled executable binaries are accepted. Scripts in particular are only a bunch of text that have no meaning for the kernel, and would lead to a runtime error.

However, Linux has a very nice feature called the shebang, which consists in writing as the first line of the script a description of the binaries executable (the interpreter) to use to read the script (eg. `#!/bin/bash`). This shebang is intercepted by the kernel on the exec entrypoint, and will invoke the described interpreter and pass the script to it. This allows you to "execute" seamlessly a script on Linux.

There is no such mechanism on Windows. The "best thing" you have is the association of a program to a kind of file to read it, the "Opens with" feature. But it is a user space functionality, not a kernel one, and its behavior depends totally on the program that implements it. Primary implemented use is when you double-click the file. Secondary implemented use is when you type the name of the file in a CMD shell. In this second situation, given that the `*.py` type of files is associated to `python.exe`, if you type:
```
.\test.py
```

The CMD shell will translated that into:
```
python.exe .\test.py
```

Which effectively run the python script.

Note that for the CMD shell, `*.bat` are the native scripts for this shell, so:
```
.\test.bat
```

Is also working out-of-the-box.

Now, what Python does of all of this in the `subprocess` module ?
* `subprocess.call(['my_executable', 'param1']) will directly pass to the `CreateProcess` method of the kernel. So as said before and whatever you do, you cannot use scripts here, only executable binaries
* `subprocess.call('my_executable param1', shell=True)` will open a CMD shell, and pass the command string to it. Here, the "Open with" feature can apply.

For several reasons, Certbot uses `subprocess.call('my_executable param1', shell=True)` to run hook scripts. So, we can have these situations:
* if you pass an executable binary, it will work directly
* if you pass a `.bat` file, it will work also, because CMD shell because BAT files are native scripts for CMD
* if you pass a `.py` file, it will work also, because Python for Windows associate `*.py` with `python.exe`, and `python.exe script.py` as run by CMD in this case, is effectively running the script
 
And finally ... `.ps1` files will not work ! Because:
* powershell scripts are not native for CMD, so it relies on the "Open with" feature
* extremely unfortunately, Notepad is the default program associated to `*.ps1` files, and so the result will be to have Notepad opened with the content of the script, as observed in this issue.
* extremely stupidly, even if you associate `powershell.exe` to `*.ps1` files, it will not work either. Indeed, on the contrary to Python interpreter where you pass the script as the first parameter, for Powershell, it is `powershell.exe -File test.ps1`. The command `powershell.exe test.ps1`, as created by CMD during the subprocess call, is invalid ...

Finally, the sane way to run hooks is what I said in the TL;DR: pass the full command of `interpreter + [whatever flags needed] + your_script`. This will work in the CMD shell, because you are passing a binary executable, and so it will work as well for the Certbot hooks.

**What can be improved ?**

Arguably passing Powershell scripts as Certbot hooks is a decent choice on Windows. The two other choices are BAT files and Python files. We need to improve Certbot hooks for Windows here, by appending the correct interpreter. This is mandatory for scripts in `renewal-hooks`, and a nice feature for the CLI hooks flags. Given the script extension, and if a `interpreter + [whatever flags needed] + your_script` is not provided on the CLI / script is in `renewal-hooks`, we would have the following behavior:
* For BAT files, it works out-of-the-box, no interpreter is needed
* For Python files, we could append the `sys.executable`, to make these scripts runnable even if Python is not installed globally and the system (and so the "Open with" is not defined for `.py` files)
* For Powershell scripts, appending `powershell.exe -File` would do the trick.

I keep the issue opened until we have a decision about what we do, or we do not. @bmw, what are your thoughts here ?

**Comment by mikenor** [comment]
In the meantime, would it be possible for somebody to edit [the forum post about the Certbot Windows beta](https://community.letsencrypt.org/t/beta-phase-of-certbot-for-windows/105822)? AFAIK that post is the closest thing to official release notes for the Windows version, and it explicitly states that .ps1 hooks work:

> Similarly to Linux, Certbot for Windows supports any file executable for this platform, if it is available in the PATH. It can be .bat, .ps1 or even executable binaries like .exe.

Obviously incorrect.

**Comment by bmw** [comment]
I updated the text to say:

> Certbot for Windows supports any file that is executable on the Windows Command Prompt, if it is available in your PATH. It’s important to note that Powershell scripts do not currently work because they cannot be run directly from the Windows Command Prompt and instead they must be given as arguments to powershell.exe. If you would like to use a Powershell script, you should probably wrap it in something like a .bat file that properly passes it to Powershell.

@mikenor, thanks for taking the time to report this issue.

@adferrand, this is a pretty unfortunate problem. I'm a little hesitant to do much mangling of the hooks given to Certbot. Introducing differences between the command/executable given to Certbot and what we actually run to me seems like it's an area ripe with potential bugs and confusion.

Trying to get into the mind of a Windows sysadmin, I wonder if they'd consider a Powershell script as an executable file. Windows seems to go out of their way to not make it one with the behavior you described above. Without us advertising that Certbot will execute `.ps1` files, would a Windows sysadmin have tried it expecting it to work? If not, I don't think we should do anything here. Given all of the things we have to work on in Certbot, I personally think one reasonable option for us is to hold off on making a change and see if others hit this issue or request a similar change.

If other people do hit this issue or we're confident that people would expect a Powershell script to work, my preference would be if we're on Windows and the file ends in `.ps1`, fix up the command to run the script with Powershell. I personally don't think we should make any other changes such as changing the Python interpreter used as I think it's just as likely if not more so that the user wants to use the system Python with their own dependencies as our own copy of Python installed specifically for Certbot.
